### PR TITLE
Bach bq sample prep

### PR DESCRIPTION
**Pull request by thijs-obj**
Prepare test of `df.get_sample()` to also work with BigQuery, but marked them all as skip-bigquery for now. The next PR (#965) will implement `get_sample()` for BigQuery and enable the tests for BigQuery too

### CODE REVIEW

File: bach/tests/functional/bach/test_df_sample.py [hunk 1]
```diff
@@ -1,15 +1,43 @@
...
 
-def test_get_sample():
-    bt = get_bt_with_test_data(True)
-    bt_sample = bt.get_sample(table_name='test_data_sample',
+def test_get_sample(engine):
+    # For reliable asserts (see below) we need more rows than the standard dataset of 11 rows has.
+    # Therefore we append the dataset to itself two times to get 33 rows in total.
+    df = get_df_with_test_data(engine, True)
+    df = df.append([df, df])
+    df = df.materialize()
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_sample.py]: done in bb28c444f90db78c7a631faeedad73f92ea73776

File: bach/tests/functional/bach/test_df_sample.py [hunk 2]
```diff
@@ -1,15 +1,43 @@
...
+    assert len(pdf_sample) > 0
+    assert len(pdf_sample) < 33
+
+
+@pytest.mark.skip_bigquery('We do not support the seed parameter for bigquery')
+def test_get_sample_seed(engine):
+    bt = get_df_with_test_data(engine, True)
+    bt_sample = bt.get_sample(table_name='test_df_sample__get_sample_seed',
                               sample_percentage=50,
                               seed=200,
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_sample.py]: good point, didn't think about that. bb28c444f90db78c7a631faeedad73f92ea73776

File: bach/tests/functional/bach/test_df_sample.py [hunk 3]
```diff
@@ -71,9 +104,18 @@ def test_sample_operations():
...
-def test_sample_operations_filter():
-    bt = get_bt_with_test_data(True)
-    bt_sample = bt.get_sample(table_name='test_data_sample',
+def _get_seed(engine) -> Optional[int]:
+    """ Return 200 if the database supports seeding. """
+    if is_bigquery(engine):
+        return None
+    if is_postgres(engine):
+        return 200
+    raise Exception()
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/tests/functional/bach/test_df_sample.py]: Moved this to the bottom in bb28c444f90db78c7a631faeedad73f92ea73776

**Comment by thijs-obj** [comment]
> Have a couple of comments and some questions:
> 
> 1. Shouldn't df be materialized before getting sample? For example:
> 
> ```python
> df['new_column'] = 'a'
> df_sample = df.get_sample(table='sampled_table', overwrite=True)
> ```
> 
> Will create the temporary table but without `new_column`.

Yes, I think you are right that this does not make a lot of sense. I changed this in the follow-up PR (not ready for review yet) https://github.com/objectiv/objectiv-analytics/pull/972


> 
> 2. I don't think we have a test for `overwrite` behavior. I'm trying to understand how it should work and I am at a point where I am not really sure if it actually works or not. And based on `DataFrame.get_sample` docstring, `"if True, the sample data is written to table_name, even if that table already exists."`, it is not working. My assumption here is that it does the temporary table based on the session but doesn't reflects the changes on the table to be overwritten. Did the following test and last `pd.testing.assert_frame_equal` should be failing:


Thanks for the elaborate test. I think you are right that this does not make sense. I think your analysis that it writes a temporary table, but does not overwrite the real table is correct. In https://github.com/objectiv/objectiv-analytics/pull/972 we'll use a permanent table, and I'll create better test cases for the overwrite behaviour.

**Comment by thijs-obj** [comment]
> I'll create better test cases for the overwrite behaviour.

The improved implementation  (https://github.com/objectiv/objectiv-analytics/pull/972, WIP) relies on a new function to create the table. The tests for this new function also test the overwrite, see behaviour https://github.com/objectiv/objectiv-analytics/pull/976

**Review by KathiaBarahona — APPROVED** [review]

**Pull request merged**